### DISCUSSION TITLE
July 25 post-incident analysis

### DISCUSSION THREAD

**Discussion by nwalters512** [PrairieLearn/PrairieLearn#10294] [Announcements]
On UTC July 25, 2024, PrairieLearn experienced an incident that resulted in degraded availability between 01:45 UTC and 03:35 UTC. We sincerely apologize for the impact of this incident, and the extended amount of time it took for us to resolve it. We wanted to share some details about the events that led to this incident, the steps we took to resolve it, and the lessons we’ve learned that will help us be more reliable and resilient in the future.

# Executive summary

Throughout the duration of the incident, users of both `us.prairielearn.com` and `ca.prairielearn.com` experienced elevated 504 errors that prevented access to certain parts of PrairieLearn, including the login page, homepage, and many instructor-facing pages. Students who were already logged in and working on an assessment were largely unaffected.

The root cause of the incident was an upgrade to a library that caused a memory leak. Misconfigured memory limits meant that the application was not automatically restarted even after becoming largely unresponsive. This resulted in PrairieLearn being unable to serve some requests for the duration of the incident.

# Timeline

- **01:45 UTC**: Users of `us.prairielearn.com` begin experiencing elevated 504 errors. These errors primarily affected visits to the login page, homepage, and instructor-facing pages. Students who were already working on assessments were largely unaffected.

- **03:10 UTC**: Users of `ca.prairielearn.com` also began experiencing the same errors.

- **03:25 UTC**: A combination of automated and user reporting alerted us to the issue.

- **03:35 UTC**: We rebooted the affected hosts, which temporarily allowed them to start serving traffic again.

- **04:35 UTC**: Based on our investigation, we identified the Sentry library upgrade to be the likely root cause and started the process of downgrading the library and validating that change.

- **05:20 UTC**: We completed a deployment of the Sentry downgrade to all environments. Subsequent monitoring showed that memory usage was once again stable.

# Background

PrairieLearn uses [Sentry](https://sentry.io/) to automatically capture and report application errors.

Around 19:00 UTC on July 24, we rolled out a new version of PrairieLearn to all environments. This included a new version of the [`@sentry/node` library](https://github.com/getsentry/sentry-javascript). In this new version, the library switched to using [OpenTelemetry](https://github.com/open-telemetry/opentelemetry-js) as the foundation on which it is built. PrairieLearn has been using OpenTelemetry on its own for application observability for several years now.

We configure Sentry with per-request ["event processors"](https://docs.sentry.io/platforms/javascript/guides/node/enriching-events/event-processors/) that enrich captured errors with request-specific information like which course and assessment the error relates to. We use Sentry's `Sentry.withScope(...)` API to isolate these event processors to each incoming request.

# Root cause

Our investigation revealed that the new version of Sentry does not properly isolate its internal state between requests when it's used with an existing OpenTelemetry setup. When we added an event processor for each request, we were actually adding it to a single, global list of event processors. So, every time we got a request, one more event processor would be stored in this list, and they would never be removed. This is a classic memory leak.

Normally, Node would eventually kill processes if they ran out of memory. However, we had inadvertently set Node's [`--max-old-space-size`](https://nodejs.org/api/cli.html#--max-old-space-sizesize-in-megabytes) option too high for the amount of memory on the EC2 instance we were running on. This means that we actually exhausted the system memory long before our Node processes reached their configured memory limit.

During further investigation, we discovered that when an EC2 instance is under extreme memory pressure, [its periodic network configuration process will fail](https://repost.aws/questions/QU-IJlEVo2Q0enTASXvagUCw/t3-micro-ec2-instance-dropped-off-network), and it will be left unable to communicate with the outside world. This resulted not only in the instance being unable to serve application traffic, but also in us being unable to connect to the instance to further investigate the issue.

All of this explains why we didn't notice the memory leak in our staging environment: the amount of memory leaked is proportional to the number of requests the application is receiving. Because our staging environment doesn't receive a lot of traffic, the memory leak wasn't evident there.

# Next steps

## Improve alerting

An hour and a half is simply too long for the site to be down without us noticing. Our existing alerting was insufficient to catch the problem in a timely manner. We plan on seeking out a better altering tool that can reach us rapidly at any time of day. We also plan to add alerts for memory consumption so that we can proactively identify similar memory leak issues in the future.

## Fine-tune Node memory limits

This problem could have been mitigated if our Node memory limits had been configured appropriately for the instance sizes we were using. Had this been the case, Node would have terminated itself after running out of memory long before we started exhausting overall system memory. This would have turned a prolonged outage into a briefer service interruption as the process would have automatically restarted in a healthy state.

## Report the underlying issue to Sentry

We consider it a bug in the new version of Sentry's JavaScript library that they fail to isolate their internal state between different requests. We've filed [a bug report](https://github.com/getsentry/sentry-javascript/issues/13055) with them that includes [an isolated reproduction of the issue we were facing](https://github.com/nwalters512/sentry-memory-leak-repro). We will not upgrade to Sentry v8 until we've either confirmed that the issue is fixed or learned how to work around it.

**Comment by firasm** [comment]
Thanks for this report here, was helpful for me to read and understand.

> An hour and a half is simply too long for the site to be down without us noticing. 

This part is also surprising to me. For one because of the default AWS tools send me bajillion notifications/emails when things are down. But mostly because I find out within minutes from students (at all hours of the day) when there is even a small problem with an assignment or the platform! I would be curious to hear (from a SWE perspective) what tool you use to report downtime :-).

**Reply by nwalters512** [reply]
> For one because of the default AWS tools send me bajillion notifications/emails when things are down.

I'm curious about your own experience here too! This does not match my experience with AWS, I've found that alerting (let alone getting emails about things) has always been a pretty manual process to set up. What kind of things do you find yourself getting AWS emails for by default?

> I would be curious to hear (from a SWE perspective) what tool you use to report downtime.

We have a ton of AWS CloudWatch alerts configured to monitor all sorts of things. Up until now, those were only piped to a Slack channel, which is too easy to miss/ignore/etc. This is why I mention "seeking out a better altering tool that can reach us rapidly at any time of day": we really just need to get a phone call when things are down to punch through do-not-disturb and such. I can safely say that this incident was a wake-up call telling us that our existing alerting strategy has to change. And it will!

> But mostly because I find out within minutes from students (at all hours of the day) when there is even a small problem with an assignment or the platform!

As do we! And during business hours, we're always watching Slack for things like this. I just missed them for the very human reason that I was cooking/eating dinner when things went south. This again speaks to the need to have "everything is on fire" alerts that can reach me no matter what.

**Comment by tdy** [comment]
> On July 26, 2024, PrairieLearn experienced an incident that resulted in degraded availability between 01:45 UTC and 03:35 UTC.

I assume you mean July 24? If so, I guess this is what happened to one of the proctored MCS exams last night.

**Reply by nwalters512** [reply]
Technically July 25 (in UTC), but yes, I was off by a day.
